kube-runtime/src/reflector/shared_store.rs

```diff
@@ -0,0 +1,401 @@
+
+    type TestSharedStore<K> = SharedStore<K, TestWatcherFactory<K>>;
+
+    impl<K> SharedStore<K, TestWatcherFactory<K>>
+    where
+        K: 'static + Resource + Clone + DeserializeOwned + Debug + Send + Sync,
+        K::DynamicType: Clone + Debug + Default + Eq + Hash + Unpin,
+    {
+        fn new(events: HashMap<(Option<String>, ListParams), Vec<Event<K>>>) -> Self {
+            Self {
+                watcher_provider: TestWatcherFactory {
+                    events: Mutex::new(events.into_iter().map(|(k, v)| (k, v.into())).collect()),
+                },
+                reflectors: HashMap::new(),
+            }
+        }
+
+        async fn spawn(self) {
+            tokio::spawn(async move {
+                self.run().for_each(|_| async {}).await;
+            });
+
+            tokio::task::yield_now().await;
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    struct TestContext {
+        reconciled: Arc<Mutex<bool>>,
+    }
+
+    impl TestContext {
+        fn new() -> Self {
+            Self {
+                reconciled: Arc::new(Mutex::new(false)),
+            }
+        }
+
+        fn reconciled(&self) -> bool {
+            *self.reconciled.lock().unwrap()
+        }
+    }
+
+    struct TestWatcherFactory<K> {
+        events: Mutex<HashMap<(Option<String>, ListParams), VecDeque<Event<K>>>>,
+    }
+
+    impl<K> CreateWatcher<K> for TestWatcherFactory<K>
+    where
+        K: 'static + Resource + Clone + DeserializeOwned + Debug + Send,
+        K::DynamicType: Hash + Eq,
+    {
+        fn all(&self, list_params: ListParams) -> BoxStream<'static, watcher::Result<Event<K>>> {
+            self.watcher(None, list_params).boxed()
+        }
+
+        fn namespaced(
+            &self,
+            namespace: &str,
+            list_params: ListParams,
+        ) -> BoxStream<'static, watcher::Result<Event<K>>>
+        where
+            K: Resource<Scope = NamespaceResourceScope>,
+        {
+            self.watcher(Some(namespace.to_string()), list_params).boxed()
+        }
+    }
+
+    impl<K> TestWatcherFactory<K>
+    where
+        K: 'static + Resource + Clone + DeserializeOwned + Debug + Send,
+        K::DynamicType: Hash + Eq,
+    {
+        fn watcher(
+            &self,
+            scope: Option<String>,
+            list_params: ListParams,
+        ) -> impl Stream<Item = watcher::Result<Event<K>>> + Send {
+            let events = self
+                .events
+                .lock()
+                .unwrap()
+                .remove(&(scope, list_params.into()))
+                .expect("There can be only one stream per ListParams");
+
+            stream::unfold(events, |mut events| async move {
+                match events.pop_front() {
+                    Some(event) => Some((Ok(event), events)),
+                    // if there's nothing left we block to simulate waiting for a change
+                    None => futures::future::pending().await,
+                }
+            })
```

Comment:
i think you can just convert the vec to a stream here and return that without doing an unfold